### PR TITLE
[pooling][forward][2D/3D] Naive implementation (#2110) plus more

### PR DESCRIPTION
**Pull request by atamazov**
This resolves issue #2110, bullet (1) and implements some related fixes.

- **[pooling]**
  - 🔴 Fixed: Removed implementation specific details from NetworkConfig (a key of InvokerCache). These details lead to correctness issues with other Solvers (like this Naive one) because of selection of wrong invoker.
    - The legacy solvers produce more Invokers than before. This leads to negligible perf drop, which unavoidable anyway.
  - **[forward]**
    - 🟢 [2D/3D] Implemented and taken into use the Naive solver.
    - 🟢 [3D] PoolingForwardNd: Implemented size checks. This enables PoolingForwardNaive for 3D when limits of PoolingForwardNd are hit.
    - 🟠 [3D] Fixed: correctness for FP16 (issue #2160) in PoolingForwardNd.
    - 🟡 [2D] Small fix: alignment of private arrays and logging.
  - **[backward]**
    - 🟡 [2D] Implemented LDS (local memory) size check. This is partial implementation of #2110 (2), which makes the solver non-applicable (instead of kernel build error).
- **[general]**
  - Added MaxLocalMemorySize to target properties
- **[tests]**
  - 🔴 Fixed: issues in test_miopendriver_regression_float and test_miopendriver_regression_half introduced in #2118.
  - [pooling]
    - 🟢 Added 2D/3D regression tests for issue #2110 (1).
    - 🟢 Added regression test for issue #2160.
    - 🟡 Enabled index verification in Full tests (--all).
    - 🟡 Introduced WORKAROUND_ISSUE_2110_2
- [NFC][driver] Added const to the arguments of GetTensor... functions.

----
[Attribution] @junliume
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/enhancement
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/bug
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/correctness
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/urgency_blocker
- Proposed reviewers:
  - @JehandadKhan
  - @zjing14

### CODE REVIEW

**Comment by atamazov** [comment]
@junliume @JehandadKhan Please let me know what do you think about:
- BF16 support in pooling
- INT8 support in pooling (we need inference only AFAIU)
- NHWC support for FP32, FP16, BF16, INT8
- The same Q about NDHWC

These features can be incorporated into the Naive solvers. What I would like to know is their importance/urgency (blocker/high/normal/low or "not needed").

File: src/solver/pooling/forwardNaive.cpp [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+namespace solver {
+
+namespace pooling {
+
+namespace {
+
+#if !MIOPEN_NDEBUG
+template <typename T>
+bool IsPower2(T v)
```
Comment (junliume) [inline src/solver/pooling/forwardNaive.cpp:45]: @atamazov  the CI is complaining about unused function template:
```
[2023-05-24T16:32:46.109Z] /home/jenkins/workspace/MLLibs_MIOpen_PR-2163/src/solver/pooling/forwardNaive.cpp:45:6: error: unused function template 'IsPower2' [-Werror,-Wunused-template]

[2023-05-24T16:32:46.109Z] bool IsPower2(T v)

[2023-05-24T16:32:46.109Z]      ^

Comment (atamazov) [reply src/solver/pooling/forwardNaive.cpp:45]: @junliume Thanks! Fixed & merged from fresh develop.

**Comment by junliume** [comment]
[WIP] CI kicked off but previously have issues, not started again.

**Review by junliume — APPROVED** [review]
Ping @JehandadKhan  for review

**Pull request merged**